tls/s2n_server_hello_retry.c

```diff
@@ -93,14 +91,18 @@ int s2n_server_hello_retry_recv(struct s2n_connection *conn)
         new_key_share_requested = (kem_group != conn->kex_params.client_kem_group_params.kem_group);
     }
 
-    /*
+    /**
      *= https://tools.ietf.org/rfc/rfc8446#section-4.1.4
      *# Clients MUST abort the handshake with an
      *# "illegal_parameter" alert if the HelloRetryRequest would not result
      *# in any change in the ClientHello.
-     */
-    POSIX_ENSURE((conn->early_data_state == S2N_EARLY_DATA_REJECTED) || new_key_share_requested,
-            S2N_ERR_INVALID_HELLO_RETRY);
+     *
+     *= https://tools.ietf.org/rfc/rfc8446#4.2.8
+     *# If either of these checks fails, then
+     *# the client MUST abort the handshake with an "illegal_parameter"
+     *# alert.
+     **/
```

Comment:
Nit, but can you swap the order of these two compliance comments? If the "If either of these checks fails" one comes first, there might be some hint that it's related to the previous compliance comments with the actual checks in them.

But even then, it doesn't seem great that these closely related checks / compliance comments are scattered all over the place. It might be a sign that the code needs to be rearranged. Can we put both checks in the same place? Maybe the key share extension?

Reply:
Since I am reusing an existing check in `s2n_server_hello_retry_recv`, which seems to me like it sort of belongs in this function, I decided to add another check for the selected group in supported groups here, so the compliance comments are together: 18f74e723658aaae9d2e0c12e57a8d4f39a998fd

If after looking at this you think it's best to refactor in a different way, let me know!